add option to not split words or children

stropho opened the pull request:
### `tokenize`

> default value: `characters`
By default, the markup is split naively by size, i.e. words can be split
To override this behaviour, you can use the following values:
 - `characters` - any text can be divided
 - `words` - each word, separated by space or other white character, is undividable entity.
  Note that a string `end.12324??!@!!abcd` does not contain any space, is considered to be a single word and cannot be split. Additionaly, you can use `&nbsp` in case you want keep some words glued together.

stropho commented:
I was just thinking about it, and I have a few points:

1. the option `children` could be misleading, and could be renamed to `leafs` - the "not having any more children" nodes
2. `words` option could actually be deleted to remove some complexity added within this PR. Users could just split their string to array of words or custom strings; and use the `leafs` option
3. It would still probably be helpful to work only with the children of root node - that is, I believe, level 2 in the code. Would that be something helpful? Or  just useless complexity? Should it be customizeable, i.e. have a `maxDepth` prop ? 

Can anyone elaborate?

patrik-piskay commented:
Thanks for the PR and suggestions @stropho! I should have time to properly review this and address the points you brought up by the end of the week.

patrik-piskay commented:
I'm seeing an issue with the "words" strategy based on looking at the generated screenshots (`__tests__/screenshots/one-line-words.png`) - it doesn't preserve whitespaces.

-----

Regarding your points:
1. I definitely want to keep the "words" option as I totally see how that option specifically could be useful so it makes sense to support it out of the box. (don't really see that with the "children" yet, happy to keep it in, but just as another option.)

2. agree that "children" is not the best name for it. I kind of like "node" more than "leaf"

3. I'd wait until that's actually requested - as you say, it adds some complexity that we don't even have a case for. Also, in case this is actually needed, I would not use a top level prop like `maxDepth` (because that's not very intuitive, i.e. _which level does `maxDepth={2}` represent? the direct children of the top level `<div>` or its grandchildren?_).
A better solution could be (_because usually users control the markup used for `<TruncateMarkup />`_) actually marking DOM nodes as not "splittable". something like this (using `truncateAtom` (_or whatever the name would be_) as a hint only, it wouldn't actually get rendered):
```
<TruncateMarkup>
  <div>
    <span truncateAtom>{ ... can have more DOM nodes inside }</span>
    <span truncateAtom>{ ... can have more DOM nodes inside }</span>
    <span truncateAtom>{ ... can have more DOM nodes inside }</span>
  </div>
</TruncateMarkup>
```
this would not split the "top level" spans' content, rendering only the ones that fully fit

---

Some feedback:
- "words" whitespace issue needs to be solved
- we need to check/validate the `tokenize` prop value - warn in console if the value used is not supported + fallback to `characters`. currently it's throwing errors
- we need more test coverage - especially for the "children" option - different depths (can be in the same test file, see `__tests__/app/src/pages/react-components.js` that tests 2 cases in the same file)

stropho commented:
Sry, for a later reply. Everyone's busy :)

1. alright, `words` seem to useful. Regarding the white-spaces and tokenizing, we could
   1. replace all white characters matching `/\s+/` by a single space. Doesn't feel right
   2. keep all white characters, and  treat all `/\s+/` as  (empty) words. Not Great Either
   3. keep white space as part of the word - that would match `/\S+\s*/`, or the first word of a sentence would be `/\s*\S+\s*/`. Personally, my favorite from these 3 options
2. and also point 3. - in my case, I need something like https://prnt.sc/mpwgx3 i.e. the text in the blue bubble can't be split. In this very specific case, I  could actually replace all spaces in the bubble by `&nbsp;` or similar and the `words` approach would solve it. However, I can imagine the sooner or later this would not be enough.

I like the approach with `<span truncateAtom>` - well, react does not support custom html attributes. At least not in all versions. Additionally, we would not want to render attributes to DOM that are basically useless. Another approach could be to use another custom component:
```
<TruncateMarkup.Atom>{ /* ... can have more DOM nodes inside */ }</TruncateMarkup.Atom>
```

Anyway, I agree that the option `children` or similar using some kind of unbreakable atom definition is not required within this PR.

One more thing I need to clarify: I'm not quite sure I understand to which case does this apply to

> * we need to check/validate the `tokenize` prop value - warn in console if the value used is not supported + fallback to `characters`. currently it's throwing errors

patrik-piskay commented:
Same here :)

Regarding the word splitting strategy, I did some testing and we should be good with doing `.split(' ')` instead of `.split(/\s+/)`. The reason for that is we are working with an HTML markup and that doesn't allow multiple spaces, tabs, etc. The only exception that should interest us is `&nbsp;`. Good thing is JSX's output for input `"Hello&nbsp;React"` is `"Hello\xA0React"` -> `\xA0` being a whitespace, but it doesn't get picked up when splitting the string by `' '`. So all non breaking spaces will be preserved.

-------

> react does not support custom html attributes
> we would not want to render attributes to DOM that are basically useless

That's actually what I originally meant by `...as a hint only, it wouldn't actually get rendered` - the idea is to check for something like `truncateAtom` when we are transforming the source React elements structure, use it as a hint that the node is no longer splittable, and remove `truncateAtom` from the element before rendering (as it did its part and is no longer needed, nor valid).

Thanks for providing an example for the children/node/leaf strategy! Let's focus on `words` in this PR and once we are happy with that, we can work on the node strategy in another one.

stropho commented:
Long time no see :)

> The reason for that is we are working with an HTML markup and that doesn't allow multiple spaces, tabs, etc.

The statement is true when we use "pure" JSX template. However, if we pass a variable to be rendered (e.g. with user input, result from DB query, etc.), all the present white character are actually in DOM (you will see them when accessing `innerHTML`). On the other hand, they are not visible for the user because of how HTML works - as you mentioned. Well, since we are splitting text that is going to be inserted to DOM, other white characters could be present. I just tried, and added in the examples. I.e. words separated by `\n` cannot be split by `.split(' ')` - delimiter `/\s+/` works

---
In case performance is an issue, here are my findings https://prnt.sc/mx9d9x :

1. best performance is when not using regexp, i.e. `.split(' ')` . I just haven't figured how to use it and at the same time put a space between words without using an additional `.map()` or similar
2. When using regexp the performance for `.split(/ /)`, `.split(/\s+/)`, or a bit more complex regexp using with `.match()` is more or less the same

Please advise what to do next. As I understand, the only remaining issue from your last post is about the white characters. Cheers

patrik-piskay commented:
I'm leaning towards using `.match(/(\s*\S+)/g)` for 2 main reasons:
1. it preserves the whitespace characters, so we should be able to support `nbsp` symbols without any difficulties
2. compared to `.match(/(\S+\s*)/g)` it captures the whitespaces _before_ words, so we don't have any trailing whitespaces to worry about when joining the words together.
example: for input string `'Hello React World'` it produces `['Hello', ' React', ' World']` instead of `['Hello ', 'React ', 'World']`, so when truncating, we can end up with `['Hello', ' React']` which is great (and not with `['Hello ', 'React ']` which is not great because of the trailing whitespace after `React`)

stropho commented:
1. I just found out, that the `&nbsp;`,i.e. in React `\xA0`, is actually matched by `/\s/`. The whole time I was working with the assumption that it is NOT matched :confounded: :astonished: and that it can be used to make multiple words unsplittable. Now, it means that the unbreakable space is actually breakable by `tokenize={words}`. So the regexp could get again a bit more complicated: let's say `.match(/(\s*[\S\xA0]+)/g)` or even `.match(/([^\S\xA0]*[\S\xA0]+)/g)` . I'll look into it this week; I actually made a mistake in the examples. Good that you have the examples mechanism in the repo :+1: 

2. I don't really see a problem with a trailing space, but I assume you have more experience with this package. If the developer wants a space between the last word and ellipses, he/she can use custom ellipses --> no problem.

patrik-piskay commented:
1. :+1:
2. `If the developer wants a space between the last word and ellipses, he/she can use custom ellipses --> no problem`     
that's actually the opposite of what I was going for. using the examples from my previous comment, the issue is that with `['Hello ', 'React ']` the trailing space is always there, and the final result is `'Hello React ...'`. So we want to have `['Hello', ' React']`, which produces `'Hello React...'` (_no space between 'React' and '...'_) - and that we'll achieve by capturing the whitespace _before_ a word, not after. Now, if a developer wants to for some reason keep the space before the ellipsis, they can do that, as you already mentioned, by defining a custom ellipsis. but the default behaviour should be no trailing space

stropho commented:
2. I understand and agree - probably I just didn't describe it clearly in my last post

3.  To make the results more clear, I created a new file to write tests for the tokenizer, i.e. regexp, or possibly other utility unit tests. I'm quite happy with the results. There is just 1 more issue that doesn't seem right. `'foo \xA0 \xA0and bar'` --> `['foo', ' \xA0 \xA0and', ' bar']`  - that would be just fine in most cases, i.e. white characters before a word will be there with the word in a single string. Another more logical solution would be to split it like this : `['foo', ' \xA0', ' \xA0and', ' bar']` , meaning that `'\xA0'` has the very same meaning as a non-white character. That could, however, lead again to a trailing space or more specifically trailing `'\xA0'` - real dilemma :expressionless: 

I hope that now it's gonna be just some final touches for the PR. If you have any more requests like renaming some vars in the code etc. , let me know

patrik-piskay commented:
Awesome, I'll take a look tomorrow!

I think your solution already works as expected - like you said, we want to avoid trailing whitespaces. It's all about the non-whitespace characters - if we can't fit anything that follows a whitespace, we shouldn't display that whitespace (or multiple whitespaces) either

patrik-piskay commented on src/index.js:
```diff
@@ -354,7 +371,17 @@ export default class TruncateMarkup extends React.Component {
       return node;
     } else if (typeof node === 'string') {
-      return this.splitString(node, splitDirections);
+      if (this.policy.tokenizeString) {
```
I think this whole block could be removed because we already have it inside the `splitString` method, and it seems that's the correct place to do it in.

stropho replied:
yes, you are right

patrik-piskay commented on src/index.js:
```diff
@@ -389,18 +416,26 @@ export default class TruncateMarkup extends React.Component {
+        this.policy.tokenizeString(string),
+        splitDirections,
+        level,
+      );
```
we need to `.join('')` after splitting otherwise the backwards compatibility breaks (because it changes the structure from a string on entry to an array on exit - React doesn't mind, but the custom ellipsis callback may have some issues with that :).

stropho replied:
ok

jasonlangdon-sky commented:
You'll need to add the tokenize prop to the index.d.ts file

jasonlangdon-sky commented:
Curious what the time frame on getting this feature merged in, I'm rather keen to make use of it (and npm installing the forked version doesn't appear to be working)

stropho commented:
I was off for a while and yesterday a bit busy. So I added the one line for typescript definition :) I hope It's ready to go.

I assume installing directly from git doesn't work because there is a build step somewhere between git and the `npm publish`

jasonlangdon-sky commented:
I tried forking your work, and making the changes (including adding a postinstall hook in the package.json file) but then ran into issues with babel not building (and I don't use Babel anywhere else) so kind of gave up. Hopefully this can be merged in soon :-)

patrik-piskay commented:
Yes, this does look good, I'll merge it in soon, but before publishing I'll need to work on more documentation and examples. Hoping for the end of this week though 🤞

patrik-piskay commented:
Released as `4.0.0`. Thanks @stropho!

stropho commented:
Just used the new version in our project and works like a charm. Now, I totally understand why you wanted me to `join('')` the words back to a single string for `ellipsis` etc. :smiley:
